src/main/webapp/script.js

```diff
@@ -1,25 +1,184 @@
+      const fieldName = this.field;
+      const startIndex = this.index;
+      this.remove();
+      updateIndices(fieldName, startIndex);
+    }
+  }
+
+  /** Gets the text in a ParameterInput's text area. */
+  get text() {
+    return this.textArea.value;
+  }
+
+  /** Gets the ParameterInput's index. */
+  get position() {
+    return this.index;
+  }
+
+  /** Gets the parent field of the ParamterInput (Tags, Ingredients, or Steps). */
+  get field() {
+    return this.parent;
+  }
+
+  /** Sets the text in a ParameterInput's text area. */
+  set text(value) {
+    this.textArea.value = value;
+  }
+
+  /** Sets the ParameterInput's index. */
+  set position(value)  {
+    this.index = parseInt(value);
+  }
+
+  /** Sets the parent field of the ParamterInput (Tags, Ingredients, or Steps). */
+  set field(value) {
+    this.parent = value;
+  }
+}
+customElements.define('parameter-input', ParameterInput);
+
+/** 
+ * Creates a ParameterInput.
+ * @param {string} name Used for displaying and determining the parent field.
+ * @param {string} index The position of the ParameterInput.
+ */
+function createParameterInput(name, index) {
+  var newParameter = document.createElement('parameter-input');
+  newParameter.setAttribute('name', name);
+  newParameter.setAttribute('index', index);
+  newParameter.setAttribute('id', name + index);
+  return newParameter;
+}
+
+/** 
+ * Inserts a ParameterInput.
+ * @param {ParameterInput} previous The ParameterInput to insert after.
+ * @param {ParameterInput} parameterInput The ParameterInput to insert.
+ */
+function insertParameterInput(previous, parameterInput) {
+  previous.insertAdjacentElement('afterend', parameterInput);
+  updateIndices(parameterInput.field, parameterInput.position + 1);
+}
+
+/** 
+ * Appends an existing ParameterInput.
+ * @param {string} fieldName The field to append to.
+ * @param {ParameterInput} parameterInput The ParameterInput to append.
+ */
+function appendParameterInput(fieldName, parameterInput) {
+  const field = document.getElementById(fieldName);
+  field.appendChild(parameterInput);
+}
+
+/** 
+ * Appends a new ParameterInput.
+ * @param {string} fieldName The field to append to.
+ */
+function appendNewParameterInput(fieldName) {
+  const field = document.getElementById(fieldName);
+  field.appendChild(createParameterInput(fieldName.slice(0, -1), field.children.length));
+}
+
+/**
+ * Updates the indices of ParameterInputs after an insertion or deletion.
+ * @param {string} fieldName The field with the ParameterInputs to update.
+ * @param {number} startIndex The index of the first ParameterInput to update.
+ */
+function updateIndices(fieldName, startIndex) {
+  var parameters = document.getElementById(fieldName).children;
+  for (var i = startIndex; i < parameters.length; i++) {
+    parameters[i].position = i;
+    parameters[i].setIndexAttributes();
+  }
+}
+
+/** Gets a parent recipe's data from Datastore. */
 function getOriginalRecipe() {
```

Comment:
Where does this get called from?

Reply:
It's called from the form that gets a spin-off and uses it to populate a recipe's data.